### ISSUE TITLE
[Tracker] Improve foliage instancing

### ISSUE THREAD

**Issue by TokisanGames** [TokisanGames/Terrain3D#43]
## Planned Improvements

* [ ] #932 
* [ ] Collision generation
* [ ] Apply transforms in mesh asset scene file
* [ ] Multi select and paint (checkbox in the lower left, or highlight)
* [ ] Change the Eye icon to MMI visibility instead of enable/disable for faster changes

------
* [ ] Drag and drop reorganize in asset dock
* [ ] Retain current height offset while sculpting like Terrain3d_objects does
* [ ] Foliage transform tool, a brush with expand/shrink scale, raise/lower, paint instance color.
* [ ] Create & destroy small MMI segments dynamically. 
* [ ] Improve generated grass: Meshless, vertex count, pointy, bend, gradient, wind
* [ ] AO/sample ground - wait for #245 

Optimizations
* Reusing RIDs
* Pre-Allocating MultiMeshes w/ maximum amount of memory
* Dynamically "create and destroy" (enable/disable/reuse preallocated buffers) MMI/MMs as the camera moves
* Updating the RS Buffer more directly with RS->multimesh_set_buffer(rid, buffer)
* Use RS->multimesh_set_visible_instances(rid, count) to limit the instances visible and the amount of the buffer used if using a preallocated maximum. Empty cells can be 0

### Biome projects
Whichever project is completed and integrated into this repo will be the most widely used.

* https://github.com/TokisanGames/Terrain3D/discussions/656
* https://github.com/jgillich/foliage3d
* https://github.com/caphindsight/Foliage3D

## Done
First pass of an MMI implementation was merged in #340, improved in #402 and individual commits

Redesign of the data structure was laid out in #524 and implemented in #523. 

* [x] Drag and drop into asset dock for meshes 99b64d768a5b5e715e92f9ccf714379a738546ec
* [x] Automatically adjust height of foliage on sculpting #424
* [x] Replace editor.cpp if instancer enable w/ add/subtract??
* [x] Option to remove all mesh types
* [x] Smaller MMI chunks 32-64/region
* [x] Store transforms in a data structure instead of an MM and generate as needed.
* [x] Store transforms in local space instead of global
* [x] Handle other meshes in scene file? - No, MMIs are for one mesh
* [x] LODs in scene files #604 
* [x] Use root node if MeshInstance3D #604 
* [x] #469 
* [x] Option to instance on all collision, not just terrain (eg grass on rocks) https://github.com/TokisanGames/Terrain3D/pull/732
* [x] Highlight MMI instances on the ground for the asset selected in the dock (temp swap override material), and/or add a picker #736 
* [x] Count used instanced #738 
* [x] Pick instance on ground to select in asset dock #734 
* [x] #788 

------------------
## Original

My idea for a paintable instancer is:
* Painting occurs on a splat/control map that can be saved in storage, but never goes in VRAM.
* MMI chunks are generated as the camera moves
* MMI chunks are regenerated if contained in the edited aabb of any terrain modifications

#324 and #278 need to be completed before this.

Reasoning:
* A control map based particle shader means placement around vertices with noise offsets. MMI nodes can be placed exactly where painted.
* Erasing instances is easy when painting on a control map. It's difficult to find and erase instances with an MMI, but regenerating a whole chunk while editing isn't a big deal.
* Particle shaders might be 2-4% faster, but will require yet another control map in VRAM. Fine if it's placement based on ground texture, but not if paintable. We're already consuming 4GB VRAM in OOTA and need to conserve.
* Particles can be frustum and distance culled. MMIs cannot, but with small enough chunks should be fine.
* MMIs seem to operate with Godot's autolod system (from what I see in SGT code). Haven't tested it, and don't know about particles.

 
--------------
## Other Notes

The current [control map format](https://terrain3d.readthedocs.io/en/latest/docs/controlmap_format.html) has 11 bits free, and most likely need 9 of them for paintable uv scale, slope, and rotation. 2 more could possibly be added to the base & overlay, or used for non-destructive layers.

Thus we probably need a second control map that could allocate foliage particles, water, and other items. However this other control map may not need to be in vram. It could be used to populate an MMI or a buffer filled with positions for a vertex shader.

Surprisingly, Godot's autolod apparently works with MMIs, according to SimpleGrassTextured, so this should be tested with MMIs and particle shaders.

### Other solutions
* [Hterrain](https://github.com/Zylann/godot_heightmap_plugin/) approach uses splatmaps to store [detail layers](https://github.com/Zylann/godot_heightmap_plugin/blob/master/addons/zylann.hterrain/hterrain_detail_layer.gd), then uses that data to generate MMI chunks

* [SimpleGrassTextured](https://github.com/IcterusGames/SimpleGrassTextured/tree/main) appears to be an MMI painter. See https://github.com/IcterusGames/SimpleGrassTextured/blob/main/addons/simplegrasstextured/grass.gd. 

* What do other solutions use? [MTerrain](https://github.com/mohsenph69/Godot-MTerrain-plugin), UE, Unity?

### My ideas
* A paintable control map Image that does not get stored in vram. This is used to generate MMI chunks. Chunks are regenerated upon sculpting updates within the editable AABB.
* Use a vram control map that feeds directly into a particle shader. Offer 2 particle types per vertex and a (paintable?) density. Several instances of the specified particles are placed around the vertex based on sampled noise and density.
* Collision could be generated on enabled particles, especially with the chunk based generator.  
* Compute based grass https://ueshita.hatenablog.jp/entry/2023/12/06/220451
* Either particle or MMI
* Should be paintable w/ multiple levels eg. paint grass, sunflowers, rocks
* Collision is not expected, but could be generated with our chunk system
* Maybe an option for the user to provide either a textured mesh, or just a 2D texture that gets applied to a generated quad

https://godotshaders.com/shader/wandering-clipmap-stylized-grass/
https://godotshaders.com/shader/wandering-clipmap-foliage-particle-process/

* Ghost of Tsushima style Grass
https://www.youtube.com/watch?v=bp7REZBV4P4&t=2s
https://www.youtube.com/watch?v=Ibe1JBF5i5Y
https://www.gdcvault.com/play/1027033/Advanced-Graphics-Summit-Procedural-Grass
Ansraer: "another neat trick is to use SDF for foliage alpha. That way you can gradually increase leaf size as objects get further away from the camera to hide the lower leaf density when it switched to a lower poly lod. If you are smart about it you could write a shader that only samples the alpha channel in the pre-pass, saving you from having to sample it in the main color pass. (due to the GPU's built in early discard feature)"
https://discord.com/channels/691957978680786944/1065519581013229578/1384197000802861057

* Consider rendering foliage without meshes
https://www.youtube.com/watch?v=R-bjXOEQyX8
https://github.com/godotengine/godot/issues/19473
https://github.com/godotengine/godot/issues/83446

* GL Grass
https://github.com/SomePineaple/GLGrass
https://www.youtube.com/watch?v=2h5NX9tIdno

* https://gpuopen.com/learn/mesh_shaders/mesh_shaders-procedural_grass_rendering/
* https://www.youtube.com/watch?v=jw00MbIJcrk

* Ghost of tsushima style grass in Godot
https://github.com/2Retr0/GodotGrass

**Comment by Saul2022** [comment]
Maybe to add up on the  description it could be  faded in the distance like the visibility ranges, and giving an ability to get replaced in the distance by another mesh/alpha  card. It could be good to see  how the roblox engine manages  a high amount of grass with a small loss on performance  with particle shader.  https://youtu.be/luhAhuR_nw4  at first it seems there's only s region   that uses grass and by the distance lods appear there.

**Comment by Saul2022** [comment]
Sounds pretty good how it coming, though something that should be taken into account are custom lods, where for example for short distance you can have mesh foliage and in longer distance a billboard mesh is used with less vertices. I think for lod system the mohsen foliage paint can be useful ( it allows various lods and are customizable if  you want to check it out https://github.com/mohsenph69/Godot-MTerrain-plugin, though don’t remember if he uses MMI). 

Also wouldn’t the  particles be even faster in 4.3 thanks to the  acyclic render graph(gpu particles)?.

**Comment by TokisanGames** [comment]
> it could be faded in the distance like the visibility ranges

Distance fade is handled in the material of the user's object. If using MMI chunks distant chunks won't be generated.

> custom lods

We'll be lucky to get any LODS at all. Custom lods with MMI chunks means 2-4x the number of chunks and instances to manage. If autolods work, and I haven't tested them yet, great. There's a proposal or idea to connect custom lods into the built in system, so that's perhaps the best way to get custom lods in MMI. However, I will look at other systems to see what they do. Links to the specific section of code are appreciated.

> Also wouldn’t the particles be even faster in 4.3 thanks to the acyclic render graph(gpu particles)?.

I don't know much about it. Do  you have a link where it talks about particles?

**Comment by Saul2022** [comment]
> > it could be faded in the distance like the visibility ranges
> 
> Distance fade is handled in the material of the user's object. If using MMI chunks distant chunks won't be generated.
> 
> > custom lods
>  I will look at other systems to see what they do. Links to the specific section of code are appreciated.
> 
Yea sounds understandable though i think for that you could talk with mohsen  about his approach, which seem’s to work well.   As for his system let’s see if i find the files where he uses it.  
 
> I don't know much about it. Do you have a link where it talks about particles?

Yes here it  is about the performance diference  when it talks results — gpu  performance.

https://godotengine.org/article/rendering-acyclic-graph/

**Comment by TokisanGames** [comment]
Good discussion on #rendering godot dev chat. Bottom line, acyclic graph won't help fixed particles like grass much. MMI is simpler and more efficient than a particle shader (which uses MMI in the background).

```
Cory
7:56 PM
I'm deciding between a Particle Shader and Multimesh Instancing for a foliage instancer in Terrain3D and have some questions that I haven't been able to find out for myself in the code:
Does the new acyclic graph improve MMI like it does particles?
Do MMIs use particles in the background?
Does the data container that stores MMI transforms consume VRAM, as a splatmap/uniform array that defines placement for a particle shader would?
It appears that MMI manages autolods, so I don't have to setup multiple MMIs to manually swap LODs. (Haven't tested it, seems to be undocumented, but SimpleGrassTextured documentation shows it working). Is that true, and does that also work on particles? https://github.com/IcterusGames/SimpleGrassTextured#lod-optimization

Any other pros and cons or recommendations on one path vs the other? I'm quite familiar with using both, but lack the esoteric knowledge of the renderer.
I'd like to support 64-128 particle types, paintable on 16k and eventually up to 90k terrains, highest instancing performance, minimal VRAM, built in LOD support if possible.

dariosamo
8:39 PM
Does the new acyclic graph improve MMI like it does particles?
No. MMI is reliant on a CPU-side structure you upload for the transforms. GPU particles benefited from the graph because the transforms are filled by the GPU instead (hence why it parallelized it better).

Cory
10:24 PM
So only the initial loading is faster with the graph, but once the particles (say grass) are in place, rendering is not any faster?
And are we talking about this speedup is for GPUParticles3D? Or a particle shader, which would likely get transforms from a texture map?

clayjohn
10:29 PM
For MMI, acyclic graph doesn't make them any faster, that's right.
The speed benefit for Particles comes from parallelization. So the speed benefits will vary depending on your content. Basically, before particle updates (so running the particles process shader) happened sequentially. After the acyclic graph they run in parallel. So if you have a lot of particle systems that don't fully saturate the GPU (i.e., there are only a few thousand particles in each GPUParticles node), then you should get a huge speedup.
If your particles use a complex shader and you have tens of thousands of particles per node, then you might not see a huge speedup

QbieShay
10:34 PM
I think MMI would be the better fit for terrain. Particles have their own compute pipeline set up
If your objects are just static then there's no benefit to the extra setup cost
I'm deciding between Particles and Multimesh Instancing for a foliage instancer in Terrain3D and have some questions that I haven't been able to find out for myself in the code:

clayjohn
10:38 PM
Following up on the rest of the questions:
Answered
No MMI's do not use particles in the background, Particles use MMI in the background. The big difference is that particles run the particle shader pass and always have color and custom in their data structure. MMI are more efficient to render as a result
Yes, the MMI buffer consumes ram. It is always the same or less than what GPUParticles consumes (especially if you are passing a splatmap as a uniform buffer)
The auto LOD for multimesh is kind of buggy. If it works for you, then that's great. HLOD is a much better fit though IMO so you can adjust the number of instances over distance as well
Like I said before MMI are more efficient to render than Particles. However, Particles are more efficient if you are doing updates to their data every frame (and those updates can be done completely on the GPU). Updating every instance transform every frame, for example, will be slow with MMI compared to using the GPU to update the instance transforms.
```

**Comment by guitarmike** [comment]
Adding collisions to objects can't come soon enough.  We can't have people running through trees!  Once that's in, I can move on from Proton Scatter.  

Love the plug-in: you're doing God's work here!

**Comment by Saul2022** [comment]
> Adding collisions to objects can't come soon enough. 

It's fine , the devs are busy with other stuff rn that is pretty important too  , in the meantime keep it with proton

**Comment by scottdavis** [comment]
Would be nice to have the ability to select more then one mesh for building forests with random trees or a field with grass and flowers

**Comment by vassembly** [comment]
> Would be nice to have the ability to select more then one mesh for building forests with random trees or a field with grass and flowers

Would be cool to have presets where each mesh has it's own weight.

Also, big thanks to all devs and contributors for making this plugin.

**Comment by Xtarsia** [comment]
Currently the way that godot applies per instance color for multimeshes, it to multiply it with the vertex color of the mesh, so that it is accessed with the COLOR built in.

This has drawbacks for any meshes using vertex color for painted weights / pivots or other techniques.

Instead, color could be applied to instance_custom instead, which would leave vertex color un modified.

**Comment by TokisanGames** [comment]
Instancer LODs is being implemented in #604.

**Comment by aidandavey** [comment]
>> Create & destroy small MMI segments dynamically

Dynamic Cells within a radius should be relatively easy to do I think.

Presumably this could use the proposed clipmap target node as a centre point or should it have a separate target?

**Comment by TokisanGames** [comment]
Mesh instances should be on the visual target. But instance collision should be on the collision target.

**Comment by Foyezes** [comment]
Smaller chunk sizes will be great. I've recently tested LOD transitions with smaller chunk sizes using ProtonScatter. I found 8x8 chunks are great for performance and LOD transition. So 2x2 to 16x16 chunk options will be good.

[Here's the LOD transition video](https://x.com/Foyezes/status/1914421428298084467)

**Comment by Saul2022** [comment]
> Smaller chunk sizes will be great. I've recently tested LOD transitions with smaller chunk sizes using ProtonScatter. I found 8x8 chunks are great for performance and LOD transition. So 2x2 to 16x16 chunk options will be good.

The issue as far as i know is more that  it will get way more messy to manage lod transition.

**Comment by Foyezes** [comment]
> > Smaller chunk sizes will be great. I've recently tested LOD transitions with smaller chunk sizes using ProtonScatter. I found 8x8 chunks are great for performance and LOD transition. So 2x2 to 16x16 chunk options will be good.
> 
> The issue as far as i know is more that it will get way more messy to manage lod transition.

How so? I feel like the performance increase is definitely worth the effort. On the post I attached, I managed to get a pretty smooth LOD transition between 3 LODs by scaling the LODs up and down based on dist. to camera. The same thing would perform worse with a larger chunk size.

**Comment by TokisanGames** [comment]
2x2 would increase the number of MMIs and draw calls by 256x. 16 by 4x. With 60+ mesh types and 3-4 lods a piece we're already at 15k nodes on a 2k section of our terrain. Going to 60k or 3.9M is going to need more care and testing.

> I feel like the performance increase is definitely worth the effort

I'd rather see hard data provided by FPS or monitors comparing apples to apples (tweaking our tool rather than comparing differently designed tools). Our instancer grid size is hard coded, but it is a constant variable and should be changeable.

**Comment by Foyezes** [comment]
> 2x2 would increase the number of MMIs and draw calls by 256x. 16 by 4x. With 60+ mesh types and 3-4 lods a piece we're already at 15k nodes on a 2k section of our terrain. Going to 60k or 3.9M is going to need more care and testing.
> 
> > I feel like the performance increase is definitely worth the effort
> 
> I'd rather see hard data provided by FPS or monitors comparing apples to apples (tweaking our tool rather than comparing differently designed tools). Our instancer grid size is hard coded, but it is a constant variable and should be changeable.

8x8 Chunk

CPU: 0.80 ms
FPS: 50
Obj: 221
Draw Calls: 180

![Image](https://github.com/user-attachments/assets/1449b4ef-3778-4ba8-bacb-e5a3f0c368cb)

![Image](https://github.com/user-attachments/assets/e7b2b622-ef3c-4583-a1c7-838f2d7aba7d)

32x32 Chunk

CPU: 0.41 ms
FPS: 46
Obj: 94
Draw Calls: 55

![Image](https://github.com/user-attachments/assets/df8dae7d-74b9-41fa-a1b1-edfd0b117982)

![Image](https://github.com/user-attachments/assets/df9779d0-5cc8-4541-83cb-bfe2d8128233)


Both scenes have ~1.2M polygons. CPU time is double on the 8x8 chunk with a 4 FPS difference. This is tested on an i5 7600 and an RX 580 2048SP.

The performance difference is slight, but I think a smaller chunk size will be better for LOD transitions and fading. Since the chunk size can't be changed after the resources have been created, there can be an option to set the chunk size beforehand.

Smaller chunks can allow for trees, as LOD switching trees with a larger chunk can be expensive and noticeable.

**Comment by TokisanGames** [comment]
Thanks for the test. The CPU time isn't a concern. GPU time is. Of course, smaller sizes would be better for LOD transition. Individual instance lod transitions would be best. But we have to balance the scale. This test of one foliage type shows 3x the draw calls. 60-256 types * 3-4 lods is a heavy load. It's something to consider adjusting or exposing in the future.

**Comment by aidandavey** [comment]
I reckon in future we can do per instance LOD plus frustum culling in a compute shader, making chunk sizes pretty much irrelevant in most cases. 

Maybe flexible cell sizes could be useful for platforms which don't support compute. It needn't be fixed at design time, we should be able to redistribute instances if the cell size were to be changed. 

I would get compute working first at a region level, then see if there is still a demand for flexible cell sizes.

**Comment by TokisanGames** [comment]
https://github.com/godotengine/godot/issues/102799 has been fixed so MMI fade mode should work. It needs to be exposed for 4.6.
